wagtail/images/forms.py

```diff
@@ -190,9 +192,44 @@ class URLGeneratorForm(forms.Form):
             ("fill", _("Resize to fill")),
         ),
     )
-    width = forms.IntegerField(label=_("Width"), min_value=0)
-    height = forms.IntegerField(label=_("Height"), min_value=0)
-    closeness = forms.IntegerField(label=_("Closeness"), min_value=0, initial=0)
+    width = forms.IntegerField(
+        label=_("Width"),
+        min_value=0,
+        widget=forms.NumberInput(
+            attrs={
+                "disabled": True,
+                "data-w-rules-target": "enable",
+                "data-w-rules": json.dumps(
+                    {"filter_method": ["width", "max", "min", "fill"]}
+                ),
+            },
+        ),
+    )
+    height = forms.IntegerField(
+        label=_("Height"),
+        min_value=0,
+        widget=forms.NumberInput(
+            attrs={
+                "disabled": True,
+                "data-w-rules-target": "enable",
+                "data-w-rules": json.dumps(
+                    {"filter_method": ["height", "max", "min", "fill"]}
+                ),
```

Comment:
I guess this is too late, but now I do wonder whether it makes more sense to put the controller on these elements instead of the form (and get the form values via [`element.form`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/form)).

The problem with hooking them as targets is that in order to use these features, you have to rely on the `<form>` having `data-controller="w-rules"` and its `data-action` predefined. In some cases it may not be easy to override the `<form>` element (e.g. in a generic template).

But I guess the difficulty in that approach is that it's harder to listen to the `change`/`keyup` events unless you attach event listeners to the form element inside the controller code (instead of using `data-action`). Also might not be a good idea to have one event listener for each element with rules...

Anyway, I think I'm overthinking this – there's only one such use case within Wagtail so far anyway. The other potential reuse is the `ImageBlock` for handling the "is decorative" checkbox, but I guess if we put the controller on the `<form>` in the page and generic templates, we should be OK...

Reply:
Thanks @laymonage - it's funny you say that, I actually did something similar in my earliest POCs.

I think for now we can leave as is, but I agree that it may be better to do go down a different track in the future. At the time though I assumed we would not want to  add global listeners and was hoping to avoid controllers adding/removing their own listeners if needed.

In saying that, it would only be a small change to support something like this, I will make this small adjustment when merging. Essentially allowing the `this.element` to be any input-like element and I will find the nearest `form`.

```html
<input type="text" name="foo" data-controller="w-rules" data-action="change@document->w-rules#resolve keyup@document->w-rules#resolve" data-w-rules-target="enable" data-w-rules='{"other": ["a", "b", "c"]}' />
```

After this, the next step would be to probably add an action filter to allow us to filter events via the current element's form. But I will not do that on this PR.


```html
<input type="text" name="foo" data-controller="w-rules" data-action="change@document->w-rules#resolve:form keyup@document->w-rules#resolve:form" data-w-rules-target="enable" data-w-rules='{"other": ["a", "b", "c"]}' />
```

Alternatively, we could just add a value to allow the listeners to be hooked up by the controller.